pulsar-io/nifi/src/main/java/org/apache/pulsar/io/nifi/NiFiSource.java

```diff
@@ -0,0 +1,188 @@
+                } catch (IOException ioe) {
+                    log.warn("Created NiFi transaction Failed", ioe);
+                }
+
+                if (null == transaction) {
+                    try {
+                        Thread.sleep(waitTimeMs);
+                    } catch (InterruptedException ioe) {
+                        log.warn("transaction could not be created, waiting and will try again " + waitTimeMs + " milliseconds.");
+                    }
+                    continue;
+                }
+
+                try {
+                    DataPacket dataPacket = transaction.receive();
+                    if (null == dataPacket) {
+                        // no data available. Wait a bit and try again
+                        try {
+                            Thread.sleep(waitTimeMs);
+                        } catch (InterruptedException ioe) {
+                            log.warn("dataPacket could not be received, waiting and will try again " + waitTimeMs + " milliseconds.");
+                        }
+                        continue;
+                    }
+
+                    final List<NiFiDataPacket> dataPackets = Lists.newArrayList();
+                    do {
+                        // Read the data into a byte array and wrap it along with the attributes into a NiFiDataPacket.
+                        final InputStream inStream = dataPacket.getData();
+                        final byte[] data = new byte[(int) dataPacket.getSize()];
+                        StreamUtils.fillBuffer(inStream, data);
+
+                        final Map<String, String> attributes = dataPacket.getAttributes();
+                        final NiFiDataPacket NiFiDataPacket = new StandardNiFiDataPacket(data, attributes);
+                        dataPackets.add(NiFiDataPacket);
+                        dataPacket = transaction.receive();
+                    } while (dataPacket != null);
+
+                    for (NiFiDataPacket dp : dataPackets) {
+                        consume(new NiFiRecord(dp));
+                    }
+
+                    // Confirm transaction to verify the data
+                    transaction.confirm();
+                    transaction.complete();
+                } catch (final IOException e) {
+                    transaction.error();
+                    log.warn("Failed to receive data from NiFi", e);
+                }
+            }
+        }
+    }
+
+    static private class NiFiRecord implements Record<NiFiDataPacket>{
+        private final NiFiDataPacket value;
+        @Getter
+        private final CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+
+        public NiFiRecord(NiFiDataPacket value) {
+            this.value = value;
+        }
+
+        @Override
+        public NiFiDataPacket getValue() {
+            return value;
+        }
+
+        @Override
+        public void ack() {
```

Comment:
you should complete the transaction here.

also you need to implement Record#fail()